unit test failure: FAIL: TestGetIDRangePath

This was seen on opensuse leap 15.0 while testing another PR. Please let me know if you have an idea to try or need more information. Thanks!

Information about the system:
```
$ uname -a
Linux vagrant-openSUSE-Leap 4.12.14-lp150.12.67-default #1 SMP Tue Jul 9 08:18:26 UTC 2019 (b610084) x86_64 x86_64 x86_64 GNU/Linux
```

```
=== RUN   TestGetIDRangePath
--- FAIL: TestGetIDRangePath (0.00s)
    fakeroot_test.go:37: unexpected error for "temporary file, uid 1 (multiple good)": could not retrieve user with UID 1: user: unknown userid 1
FAIL
coverage: 75.9% of statements
FAIL    github.com/sylabs/singularity/internal/pkg/fakeroot     0.006s
```

--- a/internal/pkg/fakeroot/fakeroot.go
+++ b/internal/pkg/fakeroot/fakeroot.go
@@ -197,7 +197,7 @@ func (c *Config) Close() error {
 // find the first available range. It doesn't return any error
 // if the user is already present and ignores the operation.
 func (c *Config) AddUser(username string) error {
-	_, err := c.GetUserEntry(username, false)
+	_, err := c.GetUserEntry(username)
 	if err == nil {
 		return nil
 	}
@@ -241,7 +241,7 @@ func (c *Config) AddUser(username string) error {
 // RemoveUser removes a user mapping entry. It returns an error
 // if there is no entry for the user.
 func (c *Config) RemoveUser(username string) error {
-	e, err := c.GetUserEntry(username, false)
+	e, err := c.GetUserEntry(username)
 	if err != nil {
 		return err
 	}
@@ -261,7 +261,7 @@ func (c *Config) RemoveUser(username string) error {
 // It returns an error if there is no entry for the user but will
 // ignore the operation if the user entry is already enabled.
 func (c *Config) EnableUser(username string) error {
-	e, err := c.GetUserEntry(username, false)
+	e, err := c.GetUserEntry(username)
 	if err != nil {
 		return err
 	}
@@ -277,7 +277,7 @@ func (c *Config) EnableUser(username string) error {
 // error if there is no entry for the user but will ignore the
 // operation if the user entry is already disabled.
 func (c *Config) DisableUser(username string) error {
-	e, err := c.GetUserEntry(username, false)
+	e, err := c.GetUserEntry(username)
 	if err != nil {
 		return err
 	}
@@ -291,7 +291,8 @@ func (c *Config) DisableUser(username string) error {
 
 // GetUserEntry returns a user entry associated to a user and returns
 // an error if there is no entry for this user.
-func (c *Config) GetUserEntry(username string, reportBadEntry bool) (*Entry, error) {
+func (c *Config) GetUserEntry(username string) (*Entry, error) {
+	var largeRangeEntries []*Entry
 	entryCount := 0
 
 	u, err := c.getUserFn(username)
@@ -305,33 +306,53 @@ func (c *Config) GetUserEntry(username string, reportBadEntry bool) (*Entry, err
 		if entry.UID == u.UID {
 			if entry.Count == validRangeCount {
 				return entry, nil
+			} else if entry.Count > validRangeCount {
+				largeRangeEntries = append(largeRangeEntries, entry)
+				continue
 			}
 			entryCount++
 		}
 	}
-	if reportBadEntry && entryCount > 0 {
+	var largestEntry *Entry
+
+	for _, e := range largeRangeEntries {
+		if largestEntry == nil {
+			largestEntry = e
+		} else if e.Count > largestEntry.Count {
+			largestEntry = e
+		}
+	}
+	if largestEntry != nil {
+		return largestEntry, nil
+	}
+
+	if entryCount > 0 {
 		return nil, fmt.Errorf(
-			"mapping entries for user %s found in %s but all with a range count different from %d",
+			"mapping entries for user %s found in %s but all with a range count lower than %d",
 			username, c.file.Name(), validRangeCount,
 		)
 	}
 	return nil, fmt.Errorf("no mapping entry found in %s for %s", c.file.Name(), username)
 }
 
+// getPwUID is also used for mocking purpose
+var getPwUID = user.GetPwUID
+var getPwNam = user.GetPwNam
+
 // GetIDRange determines UID/GID mappings based on configuration
 // file provided in path.
 func GetIDRange(path string, uid uint32) (*specs.LinuxIDMapping, error) {
-	config, err := GetConfig(path, false, nil)
+	config, err := GetConfig(path, false, getPwNam)
 	if err != nil {
 		return nil, err
 	}
 	defer config.Close()
 
-	userinfo, err := user.GetPwUID(uid)
+	userinfo, err := getPwUID(uid)
 	if err != nil {
 		return nil, fmt.Errorf("could not retrieve user with UID %d: %s", uid, err)
 	}
-	e, err := config.GetUserEntry(userinfo.Name, true)
+	e, err := config.GetUserEntry(userinfo.Name)
 	if err != nil {
 		return nil, err
 	}

--- a/internal/pkg/fakeroot/fakeroot_test.go
+++ b/internal/pkg/fakeroot/fakeroot_test.go
@@ -26,6 +26,67 @@ type set struct {
 	expectedMapping *specs.LinuxIDMapping
 }
 
+var users = map[uint32]user.User{
+	0: {
+		Name:  "root",
+		UID:   0,
+		GID:   0,
+		Dir:   "/root",
+		Shell: "/bin/sh",
+	},
+	1: {
+		Name:  "daemon",
+		UID:   1,
+		GID:   1,
+		Dir:   "/usr/sbin",
+		Shell: "/usr/sbin/nologin",
+	},
+	2: {
+		Name:  "bin",
+		UID:   2,
+		GID:   2,
+		Dir:   "/bin",
+		Shell: "/usr/sbin/nologin",
+	},
+	3: {
+		Name:  "sys",
+		UID:   3,
+		GID:   3,
+		Dir:   "/dev",
+		Shell: "/usr/sbin/nologin",
+	},
+	4: {
+		Name:  "sync",
+		UID:   4,
+		GID:   4,
+		Dir:   "/bin",
+		Shell: "/usr/sbin/nologin",
+	},
+	5: {
+		Name:  "games",
+		UID:   5,
+		GID:   5,
+		Dir:   "/bin",
+		Shell: "/usr/sbin/nologin",
+	},
+}
+
+func getPwNamMock(username string) (*user.User, error) {
+	for _, u := range users {
+		if u.Name == username {
+			return &u, nil
+		}
+	}
+	return nil, fmt.Errorf("no user found for %s", username)
+}
+
+func getPwUIDMock(uid uint32) (*user.User, error) {
+	if u, ok := users[uid]; ok {
+		return &u, nil
+	}
+	return nil, fmt.Errorf("no user found with ID %d", uid)
+}
+
 func testGetIDRange(t *testing.T, s set) {
 	idRange, err := GetIDRange(s.path, s.uid)
 	if err != nil && s.expectedMapping != nil {
@@ -49,6 +110,14 @@ func TestGetIDRangePath(t *testing.T) {
 	test.DropPrivilege(t)
 	defer test.ResetPrivilege(t)
 
+	// mock user database (https://github.com/sylabs/singularity/issues/3957)
+	getPwUID = getPwUIDMock
+	getPwNam = getPwNamMock
+	defer func() {
+		getPwUID = user.GetPwUID
+		getPwNam = user.GetPwNam
+	}()
+
 	f, err := fs.MakeTmpFile("", "subid-", 0700)
 	if err != nil {
 		t.Fatalf("failed to create temporary file")
@@ -58,10 +127,13 @@ func TestGetIDRangePath(t *testing.T) {
 	var subIDContent = `
 root:100000:65536
 1:165536:1
+1:165536:165536
 1:165536:65536
 2:2000000:-1
 3:-1:65536
 4:2065536:1
+5:5065536:131072
+5:5065536:1000000
 `
 
 	f.WriteString(subIDContent)
@@ -117,6 +189,16 @@ root:100000:65536
 			name: "temporary file, uid 4 (multiple bad)",
 			path: f.Name(),
 			uid:  4,
+		},
+		{
+			name: "temporary file, uid 5 (multiple large)",
+			path: f.Name(),
+			uid:  5,
+			expectedMapping: &specs.LinuxIDMapping{
+				ContainerID: 1,
+				HostID:      5065536,
+				Size:        1000000,
+			},
 		},
 		{
 			name: "temporary file, uid 8 (doesn't exist)",
@@ -206,74 +288,63 @@ func createConfig(t *testing.T) string {
 
 func testGetUserEntry(t *testing.T, config *Config) {
 	tests := []struct {
-		desc           string
-		username       string
-		reportBadEntry bool
-		expectSuccess  bool
+		desc          string
+		username      string
+		expectSuccess bool
 	}{
 		{
-			desc:           "ValidUser",
-			username:       "valid_10",
-			reportBadEntry: false,
-			expectSuccess:  true,
+			desc:          "ValidUser",
+			username:      "valid_10",
+			expectSuccess: true,
 		},
 		{
-			desc:           "ValidUserReportBadEntry",
-			username:       "valid_10",
-			reportBadEntry: true,
-			expectSuccess:  true,
+			desc:          "ValidUserReportBadEntry",
+			username:      "valid_10",
+			expectSuccess: true,
 		},
 		{
-			desc:           "NoUser",
-			username:       "nouser_10",
-			reportBadEntry: false,
-			expectSuccess:  false,
+			desc:          "NoUser",
+			username:      "nouser_10",
+			expectSuccess: false,
 		},
 		{
-			desc:           "NoUserReportBadEntry",
-			username:       "nouser_10",
-			reportBadEntry: true,
-			expectSuccess:  false,
+			desc:          "NoUserReportBadEntry",
+			username:      "nouser_10",
+			expectSuccess: false,
 		},
 		{
-			desc:           "BadStartUser",
-			username:       "badstart_20",
-			reportBadEntry: false,
-			expectSuccess:  false,
+			desc:          "BadStartUser",
+			username:      "badstart_20",
+			expectSuccess: false,
 		},
 		{
-			desc:           "BadStartUserReportBadEntry",
-			username:       "badstart_20",
-			reportBadEntry: true,
-			expectSuccess:  false,
+			desc:          "BadStartUserReportBadEntry",
+			username:      "badstart_20",
+			expectSuccess: false,
 		},
 		{
-			desc:           "DisabledUser",
-			username:       "disabled_40",
-			reportBadEntry: false,
-			expectSuccess:  true,
+			desc:          "DisabledUser",
+			username:      "disabled_40",
+			expectSuccess: true,
 		},
 		{
-			desc:           "DisabledUserReportBadEntry",
-			username:       "disabled_40",
-			reportBadEntry: true,
-			expectSuccess:  true,
+			desc:          "DisabledUserReportBadEntry",
+			username:      "disabled_40",
+			expectSuccess: true,
 		},
 		{
-			desc:           "SameUser",
-			username:       "sameuser_50",
-			reportBadEntry: false,
-			expectSuccess:  false,
+			desc:          "SameUser",
+			username:      "sameuser_50",
+			expectSuccess: false,
 		},
 		{
-			desc:           "SameUserReportBadEntry",
-			username:       "sameuser_50",
-			reportBadEntry: true,
-			expectSuccess:  false,
+			desc:          "SameUserReportBadEntry",
+			username:      "sameuser_50",
+			expectSuccess: false,
 		},
 	}
 	for _, tt := range tests {
-		_, err := config.GetUserEntry(tt.username, tt.reportBadEntry)
+		_, err := config.GetUserEntry(tt.username)
 		if err != nil && tt.expectSuccess {
 			t.Errorf("unexpected error for %q: %s", tt.desc, err)
 		} else if err == nil && !tt.expectSuccess {
@@ -395,11 +466,11 @@ func testEditEntry(t *testing.T, config *Config) {
 	defer config.Close()
 
 	// this entry was removed
-	if _, err := config.GetUserEntry("valid_10", false); err == nil {
+	if _, err := config.GetUserEntry("valid_10"); err == nil {
 		t.Errorf("unexpected entry found for valid_10 user")
 	}
 	// this entry was disabled
-	e, err := config.GetUserEntry("valid_11", false)
+	e, err := config.GetUserEntry("valid_11")
 	if err != nil {
 		t.Errorf("unexpected error for valid_11 user")
 	}
@@ -407,7 +478,7 @@ func testEditEntry(t *testing.T, config *Config) {
 		t.Errorf("valid_11 user entry should be disabled")
 	}
 	// this entry was enabled
-	e, err = config.GetUserEntry("disabled_40", false)
+	e, err = config.GetUserEntry("disabled_40")
 	if err != nil {
 		t.Errorf("unexpected error for disabled_40 user")
 	}
@@ -416,7 +487,7 @@ func testEditEntry(t *testing.T, config *Config) {
 	}
 	// this entry was added and range start should be
 	// equal to startMax (as it replace valid_10)
-	e, err = config.GetUserEntry("valid_21", false)
+	e, err = config.GetUserEntry("valid_21")
 	if err != nil {
 		t.Errorf("unexpected error for valid_21 user")
 	}

